crates/sui-core/src/execution_engine.rs

```diff
@@ -470,6 +535,121 @@ fn pay<S>(
     Ok(())
 }
 
+fn update_input_coins<S>(
+    temporary_store: &mut TemporaryStore<S>,
+    coin_objects: &mut Vec<Object>,
+    gas_coin: &Coin,
+    recipient: Option<SuiAddress>,
+) {
+    let mut gas_coin_obj = coin_objects.remove(0);
+    // unwrap is safe because we checked that it was a coin object above.
+    // update_contents_without_version_change b/c this is the gas coin,
+    // whose version will be bumped upon gas payment.
+    gas_coin_obj
+        .data
+        .try_as_move_mut()
+        .unwrap()
+        .update_contents_without_version_change(
+            bcs::to_bytes(gas_coin).expect("Coin serialization should not fail"),
+        );
+    if let Some(recipient) = recipient {
+        gas_coin_obj.transfer_without_version_change(recipient);
+    }
+    temporary_store.write_object(gas_coin_obj, WriteKind::Mutate);
+
+    for coin_object in coin_objects.iter() {
+        temporary_store.delete_object(&coin_object.id(), coin_object.version(), DeleteKind::Normal)
+    }
+}
+
+fn pay_sui<S>(
+    temporary_store: &mut TemporaryStore<S>,
+    coin_objects: &mut Vec<Object>,
+    recipients: Vec<SuiAddress>,
+    amounts: Vec<u64>,
+    tx_ctx: &mut TxContext,
+) -> Result<(), ExecutionError> {
+    let (mut coins, coin_type) = check_coins(coin_objects)?;
+    if coin_type != GasCoin::type_() {
+        return Err(ExecutionError::new_with_source(
+            ExecutionErrorKind::NonSuiCoin,
+            format!(
+                "PaySui transaction requires SUI coin, {} was found",
+                coin_type
+            ),
+        ));
+    };
+    check_recipients(&recipients, &amounts)?;
+    let (total_coins, total_amount) = check_total_coins(&coins, &amounts)?;
+    let mut remaining_coins = total_coins;
+
+    for (recipient, amount) in recipients.iter().zip(amounts) {
```

Comment:
The way the rest of this function is implemented is a bit fragile. I think a clean interface is the following:
1. We define a few APIs in coin.rs for Coin object: merge, split, transfer.
2. In the implementation of PaySui, we first merge all coins, and then for each recipient we call split; for PayAllSui, we first merge all coins, and then transfer.

This makes it a lot easier to audit the code in terms of balance consistency